### PR TITLE
[processing] Fix save vector features to file algorithm's handling of pre-existing files

### LINKED ISSUE #54333: Save features to files overwrites whole GeoPackage
### What is the bug or the crash?

A pre-existing GeoPackage gets overwritten by QGIS “Save features to files”

### Steps to reproduce the issue

Save a vector layer to a GeoPackage using QGIS “Save vector features to file”

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.28.9-Firenze | QGIS code branch | Release 3.28
-- | -- | -- | --
Qt version | 5.15.10
Python version | 3.11.4
GDAL/OGR version | 3.7.1
PROJ version | 9.2.1
EPSG Registry database version | v10.088 (2023-05-13)
GEOS version | 3.12.0-CAPI-1.18.0
SQLite version | 3.42.0
PDAL version | 2.5.5
PostgreSQL client version | unknown
SpatiaLite version | 5.1.0
QWT version | 6.1.6
QScintilla2 version | 2.13.3
OS version | macOS 13.4
  |   |   |  
Active Python plugins
QuickOSM | 2.2.3
profiletool | 4.2.6
processing_saga_nextgen | 0.0.7
Serval | 3.10.5
valuetool | 3.0.17
processing | 2.12.99
sagaprovider | 2.12.99
grassprovider | 2.12.99
db_manager | 0.1.20
MetaSearch | 0.3.6

</body></html>QGIS version
3.28.9-Firenze
QGIS code branch
[Release 3.28](https://github.com/qgis/QGIS/tree/release-3_28)
Qt version
5.15.10
Python version
3.11.4
GDAL/OGR version
3.7.1
PROJ version
9.2.1
EPSG Registry database version
v10.088 (2023-05-13)
GEOS version
3.12.0-CAPI-1.18.0
SQLite version
3.42.0
PDAL version
2.5.5
PostgreSQL client version
unknown
SpatiaLite version
5.1.0
QWT version
6.1.6
QScintilla2 version
2.13.3
OS version
macOS 13.4

Active Python plugins
QuickOSM
2.2.3
profiletool
4.2.6
processing_saga_nextgen
0.0.7
Serval
3.10.5
valuetool
3.0.17
processing
2.12.99
sagaprovider
2.12.99
grassprovider
2.12.99
db_manager
0.1.20
MetaSearch
0.3.6

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

The main issue here is that on macOS, when dealing with GeoPackages, one is _always_ asked if one wants to fully replace the pre-existing GeoPackage.
Some tools will do that, some other tools will add the layers to the GeoPackage ; but it's impossible to know in advance.
And it's not clear from the description of “Save vector features to files” that it is intended to overwrite.
The “Package layers” tool for examples has a different behaviour.

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR harmonize handling of preexisting files by the save vector features to file algorithm by informing users of default action (file overwrite) and allow to switch to other action/behavior (keep file, overwrite layer, keep file and layer, etc.).

image

Fixes #54333 .

### CODE REVIEW

**Comment by DelazJ** [comment]
@nirvn the [not selected] entry in the screenshot... is a leftover from previous design? It is not an option, right?

File: src/analysis/processing/qgsalgorithmsavefeatures.cpp [hunk 1]
```diff
@@ -122,24 +161,24 @@ QVariantMap QgsSaveFeaturesAlgorithm::processAlgorithm( const QVariantMap &param
...
       layerName = match.captured( 1 );
     }
-    filePath = dest.mid( 0, separatorIndex );
+    filePath = destination.mid( 0, separatorIndex );
   }
 
   QVariantMap outputs;
-  outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+  outputs.insert( QStringLiteral( "OUTPUT" ), destination );
   outputs.insert( QStringLiteral( "FILE_PATH" ), filePath );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsavefeatures.cpp:180]:
```suggestion
  outputs.insert( QStringLiteral( "FILE_PATH" ), finalFileName );
```

Comment (nirvn) [reply src/analysis/processing/qgsalgorithmsavefeatures.cpp]: @nyalldawson , addressed.

**Comment by nirvn** [comment]
@DelazJ , it was an "option" until just now. The "[Not selected]" item is automatically added as the first option of an optional enum parameter. That's a hard-coded QGIS processing behavior.

That said, I've just made the enum non-optional after a discussion with Nyall. While we needed the enum parameter to be optional to avoid API breakage, other forthcoming changes in the way QGIS processing handles things will improve things on that front.

**Review by nyalldawson — APPROVED** [review]

**Comment by dlejay** [comment]
@nirvn Why have you chosen “overwrite file” as default behaviour ?

**Comment by nirvn** [comment]
@dlejay , to retain previously established behavior as the default one.

**Comment by agiudiceandrea** [comment]
Hi @nirvn, please see https://github.com/qgis/QGIS/issues/56517.

**Pull request merged**